### DISCUSSION TITLE
Binary Data from .ook

### DISCUSSION THREAD

**Discussion by Crsarmv7l** [merbanan/rtl_433#2490] [Q&A] [answered]
I have an .ook file that I would like to be able to extract the binary data from, and would like to understand the process so I can apply it to other files.

I have referred to this, as well as read the whole primer many times:
https://triq.org/rtl_433/IQ_FORMATS.html#i-q-sample-data-formats

as well as searched the repo for .ook and read the relevant .h header files.

I understand that the header of the .ook format contains informational data, and that the actual data contains a number for pulse duration and another for gap duration.

I assume you look at the data to figure out the type of signal?

Using the as an example:
`;pulse data
;version 1
;timescale 1us
;created 2023-03-07 14:29:31-0500
;ook 369 pulses
;freq1 433911648
;centerfreq 433920000 Hz
;samplerate 250000 Hz
;sampledepth 8 bits
;range 42.1 dB
;rssi -0.1 dB
;snr 21.4 dB
;noise -21.5 dB
548 996
500 992
500 1004
492 1976
496 1972
504 1972
500 1980
488 1004
524 1972
496 1000
500 992
504 1000
488 996
504 1004
496 988
504 1004
488 1972
504 1000
488 1976
500 1972
504 1976
496 984`

I see that the pulse durations are all roughly 500, with the exception of the first pulse at 548, while the gap durations seem to vary. It kind of looks like PPM (this sample is from Nexus which is PPM) . But I was hoping for a little clarity and maybe an explanation ?

Thanks for any help

**Comment by zuckschwerdt** [comment, marked as answer]
The concept you are looking for is "bit slicer". Basically you need to make some assumptions about the signal (e.g. PPM), recover the timing (e.g. each pulse is a clock) and collect resulting bits.

**Reply by Crsarmv7l** [reply]
Thanks, that jives with what I suspected.

**Comment by Crsarmv7l** [comment]
One more question. How is freq1 calculated for the .ook file?

**Reply by zuckschwerdt** [reply]
The estimated signal frequency is an average over the [instant phase angle](https://en.wikipedia.org/wiki/Instantaneous_phase_and_frequency).

**Reply by Crsarmv7l** [reply]
Awesome, thank you again
